compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/ForLoopsLowering.kt

```diff
@@ -178,18 +178,20 @@ private class RangeLoopTransformer(
             return super.visitBlock(expression)  // Not a for-loop block.
         }
 
-        with(expression.statements) {
+        val returnExpression = super.visitBlock(expression) as IrBlock
```

Comment:
> Is this commit just a refactoring/optimization, or was there a case that wasn't being handled correctly?

Changing the order is needed for BCE in order to replace operators in nested loops if they use induction variable from outer loop. The previous order didn't allow to do this because nested loops use iterators while we optimise outer one.

Reply:
I gave this some more thought, and I actually don't know what is the case you are referring to. Do you have an example where the order matters? If so, could you add a test for the example as well? If there is a case where the order matters, let's make sure to test it.

If the inner loop is using the induction variable from the outer loop for array access, I don't understand why it matters that the inner loop is still using iterators. If you have the "main loop variable" that is safe for array access from the outer loop, you should be able to replace all uses of that variable (for array access) in the inner loop first, or later on.

Everything else in this PR looks good, I'm just trying to understand this particular change, because there may be other similar cases to consider, or unintended consequences.

Reply:
I try to remember and failed. I checked now some cases and I think that you are right. May be I fixed with this change some other problems that were fixed later with proper way. So I reverted changing of the order. 
Moreover while I was thinking about this, I have found one more problem and fixed it. Thank you!